# MovieLens with SAP HANA PAL APRIORI


## **Initialize the connection**

In [1]:
import sqlalchemy, os
from sqlalchemy import create_engine

%reload_ext sql
%config SqlMagic.displaylimit = 5
%config SqlMagic.feedback = False
%config SqlMagic.autopandas = True

hxe_connection = 'hana://ML_USER:Welcome18@hxehost:39015';

%sql $hxe_connection

import pandas as pd
pd.options.display.max_rows = 100

## **Cleanup Model Tables and APL objects**

In [2]:
%%sql
-- --------------------------------------------------------------------------
-- drop function in/out tables, helper tables and views
-- --------------------------------------------------------------------------
drop table pal_movielens_parameters;
drop table pal_movielens_apriori_pmmlmodel;
drop table pal_movielens_apriori_result;
drop  view pal_movielens_apriori_data_input;

 * hana://ML_USER:Welcome18@hxehost:39015


Empty DataFrame
Columns: []
Index: []

## **Create Model Specific table **

In [3]:
%%sql
-- --------------------------------------------------------------------------
-- create the config and output tables
-- --------------------------------------------------------------------------
create column table pal_movielens_parameters (
	param_name   varchar(100),
	int_value    integer,
	double_value double,
	string_value varchar (100)
);

create column table pal_movielens_apriori_result (
	prerule    varchar(500),
	postrule   varchar(500),
	support    double,
	confidence double,
	lift       double
);
create row table pal_movielens_apriori_pmmlmodel (
  row_index integer,
  model_content clob
);
-- --------------------------------------------------------------------------
-- create the input data view
-- --------------------------------------------------------------------------
create view pal_movielens_apriori_data_input as
select userid, movieid
from   movielens_ratings;

 * hana://ML_USER:Welcome18@hxehost:39015


Empty DataFrame
Columns: []
Index: []

## **Run the Recommendation algorithm**

In [4]:
%%sql
-- --------------------------------------------------------------------------
-- configuration
-- --------------------------------------------------------------------------
truncate table pal_movielens_parameters;
insert into pal_movielens_parameters values ('MIN_SUPPORT'   , null, 0.1 , null); -- no default
insert into pal_movielens_parameters values ('MIN_CONFIDENCE', null, 0.1 , null); -- no default
insert into pal_movielens_parameters values ('MIN_LIFT'      , null, 0.0 , null); -- default is 0.0
insert into pal_movielens_parameters values ('MAX_CONSEQUENT', 1   , null, null); -- default is 500
insert into pal_movielens_parameters values ('MAXITEMLENGTH' , 2   , null, null); -- default is 5
insert into pal_movielens_parameters values ('UBIQUITOUS'    , null, 1.0 , null); -- default is 1.0
insert into pal_movielens_parameters values ('PMML_EXPORT'   , 1   , null, null); -- default is 0
insert into pal_movielens_parameters values ('TIMEOUT'       , 3600, null, null); -- default is 3600
insert into pal_movielens_parameters values ('THREAD_RATIO'  , null, 0.0 , null); -- default is 0.0
select * from pal_movielens_parameters;

 * hana://ML_USER:Welcome18@hxehost:39015


param_name  int_value  double_value string_value
0     MIN_SUPPORT        NaN           0.1         None
1  MIN_CONFIDENCE        NaN           0.1         None
2        MIN_LIFT        NaN           0.0         None
3  MAX_CONSEQUENT        1.0           NaN         None
4   MAXITEMLENGTH        2.0           NaN         None
5      UBIQUITOUS        NaN           1.0         None
6     PMML_EXPORT        1.0           NaN         None
7         TIMEOUT     3600.0           NaN         None
8    THREAD_RATIO        NaN           0.0         None

In [5]:
%%sql
truncate table pal_movielens_apriori_result;
truncate table pal_movielens_apriori_pmmlmodel;
-- --------------------------------------------------------------------------
-- execute the pal function to train the model
-- --------------------------------------------------------------------------
call _sys_afl.pal_apriori(
    pal_movielens_apriori_data_input
  , pal_movielens_parameters
  , pal_movielens_apriori_result
  , pal_movielens_apriori_pmmlmodel
) with overview;

 * hana://ML_USER:Welcome18@hxehost:39015


variable                                        table
0       P3     "ML_USER"."PAL_MOVIELENS_APRIORI_RESULT"
1       P4  "ML_USER"."PAL_MOVIELENS_APRIORI_PMMLMODEL"

## **Check the output**

In [6]:
%sql select * from pal_movielens_apriori_result;

 * hana://ML_USER:Welcome18@hxehost:39015


prerule postrule   support  confidence      lift
0        150      208  0.123696    0.415000  2.464292
1        208      150  0.123696    0.734513  2.464292
2        165      208  0.111773    0.528169  3.136296
3        208      165  0.111773    0.663717  3.136296
4        592      208  0.128167    0.438776  2.605472
5        208      592  0.128167    0.761062  2.605472
6        356      208  0.144560    0.284457  1.689124
7        208      356  0.144560    0.858407  1.689124
8        593      208  0.125186    0.276316  1.640778
9        208      593  0.125186    0.743363  1.640778
10       589      208  0.119225    0.337553  2.004406
11       208      589  0.119225    0.707965  2.004406
12       588      208  0.120715    0.376744  2.237127
13       208      588  0.120715    0.716814  2.237127
14       457      208  0.131148    0.413146  2.453280
15       208      457  0.131148    0.778761  2.453280
16       377      208  0.114754    0.427778  2.540167
17       208      377  0.114754    0.681416  2.540167
18       367      208  0.114754    0.490446  2.912294
19       208      367  0.114754    0.681416  2.912294
20        47      208  0.111773    0.373134  2.215691
21       208       47  0.111773    0.663717  2.215691
22       110      208  0.128167    0.377193  2.239792
23       208      110  0.128167    0.761062  2.239792
24       480      208  0.143070    0.350365  2.080486
25       208      480  0.143070    0.849558  2.080486
26       500      208  0.108793    0.477124  2.833189
27       208      500  0.108793    0.646018  2.833189
28       153      208  0.104322    0.542636  3.222199
29       208      153  0.104322    0.619469  3.222199
30       296      208  0.143070    0.296296  1.759423
31       208      296  0.143070    0.849558  1.759423
32       364      208  0.113264    0.380000  2.256460
33       208      364  0.113264    0.672566  2.256460
34       590      208  0.138599    0.460396  2.733856
35       208      590  0.138599    0.823009  2.733856
36       318      208  0.110283    0.237942  1.412913
37       208      318  0.110283    0.654867  1.412913
38       595      208  0.107303    0.409091  2.429204
39       208      595  0.107303    0.637168  2.429204
40       380      208  0.137109    0.464646  2.759095
41       208      380  0.137109    0.814159  2.759095
42       344      208  0.117735    0.451429  2.680607
43       208      344  0.117735    0.699115  2.680607
44       231      208  0.108793    0.462025  2.743531
45       208      231  0.108793    0.646018  2.743531
46       597      208  0.111773    0.510204  3.029619
47       208      597  0.111773    0.663717  3.029619
48       316      208  0.110283    0.510345  3.030455
49       208      316  0.110283    0.654867  3.030455
...      ...      ...       ...         ...       ...
6974    3793     3578  0.123696    0.653543  2.723774
6975    3578     3793  0.123696    0.515528  2.723774
6976    6539     3578  0.132638    0.631206  2.630677
6977    3578     6539  0.132638    0.552795  2.630677
6978    8961     3578  0.101341    0.539683  2.249236
6979    3578     8961  0.101341    0.422360  2.249236
6980     608     1617  0.126677    0.379464  2.036964
6981    1617      608  0.126677    0.680000  2.036964
6982    6539    33794  0.107303    0.510638  3.263222
6983   33794     6539  0.107303    0.685714  3.263222
6984    8961    33794  0.107303    0.571429  3.651701
6985   33794     8961  0.107303    0.685714  3.651701
6986    4226     5445  0.101341    0.515152  3.005797
6987    5445     4226  0.101341    0.591304  3.005797
6988    6539     5445  0.105812    0.503546  2.938082
6989    5445     6539  0.105812    0.617391  2.938082
6990    6539     4896  0.114754    0.546099  3.489835
6991    4896     6539  0.114754    0.733333  3.489835
6992    6539     5989  0.105812    0.503546  3.412924
6993    5989     6539  0.105812    0.717172  3.412924
6994    6539     4226  0.102832    0.489362  2.487589
6995    4226     6539  0.102832    0.522727  2.487589
6996    6

In [7]:
%sql select * from pal_movielens_apriori_pmmlmodel;

 * hana://ML_USER:Welcome18@hxehost:39015


row_index                                      model_content
0          1  <PMML version="4.0" xmlns="http://www.dmg.org/...

## **Collaborative filtering - Create the result view**

In [8]:
%sql drop   view pal_movielens_apriori_result_collaborative;

 * hana://ML_USER:Welcome18@hxehost:39015


Empty DataFrame
Columns: []
Index: []

In [9]:
%%sql
create view pal_movielens_apriori_result_collaborative as
select *
from (
  select
      t1.userid
    , row_number() over(partition by t1.userid order by t1.score desc, t1.consequent desc ) as rank
    , t1.consequent as movieid
    , t1.score
    , movies.title
    , movies.genres
    , links.imdbid
    , links.tmdbid
  from (
    select input_data.userid, rules.postrule as consequent, max(rules.confidence) as score
    from movielens_ratings as input_data
    left outer join (select * from pal_movielens_apriori_result) rules on (cast (input_data.movieid as varchar(500)) = rules.prerule)
    where rules.postrule is not null
    group by input_data.userid, rules.postrule
  ) t1
    left outer join movielens_movies movies on movies.movieid = t1.consequent
    left outer join movielens_links  links  on links.movieid  = t1.consequent
) t1
where t1.rank <= 5;

 * hana://ML_USER:Welcome18@hxehost:39015


Empty DataFrame
Columns: []
Index: []

## **Content-based filtering - Create the result view**

In [10]:
%sql drop   view pal_movielens_apriori_result_contentbased;

 * hana://ML_USER:Welcome18@hxehost:39015


Empty DataFrame
Columns: []
Index: []

In [11]:
%%sql
create view pal_movielens_apriori_result_contentbased as
select *
from (
  select
      t1.movieid
    , row_number() over(partition by t1.movieid order by t1.score desc, t1.consequent desc ) as rank
    , t1.consequent as similar_movieid
    , t1.score
    , movies.title
    , movies.genres
    , links.imdbid
    , links.tmdbid
  from (
    select movieid, rules.postrule as consequent, rules.confidence as score
    from movielens_movies as input_data
    left outer join (select * from pal_movielens_apriori_result) rules on (cast (input_data.movieid as varchar(500)) = rules.prerule)
    where rules.postrule is not null
  ) t1
    left outer join movielens_movies movies on movies.movieid = t1.consequent
    left outer join movielens_links  links  on links.movieid  = t1.consequent
) t1
where t1.rank <= 5;

 * hana://ML_USER:Welcome18@hxehost:39015


Empty DataFrame
Columns: []
Index: []

## **Collaborative filtering - Validate the result**
Let's verify how many users will actually get recommendations using the following SQL:

In [12]:
%%sql
select reco_count, count(1) as user_count
from (
  select userid, max(rank) as reco_count
  from   pal_movielens_apriori_result_collaborative
  group by userid
) group by reco_count order by reco_count desc;

 * hana://ML_USER:Welcome18@hxehost:39015


reco_count  user_count
0           5         660

Let's verify how many distinct movies will actually get recommended to a user (part of the top 5 scores) using the following SQL:

In [13]:
%%sql
select
    count(1) as movie_count
  , count(1) *100 / (select count(1) as count from movielens_movies ) as movie_ratio
from (
  select movieid
  from   pal_movielens_apriori_result_collaborative
  group by movieid
);

 * hana://ML_USER:Welcome18@hxehost:39015


movie_count movie_ratio
0           41    0.449315

Let's verify how many distinct movies will potentially get recommended to a user (not just the top 5 scores) using the following SQL:

In [14]:
%%sql
select
    count(1) as movie_count
  , count(1) *100 / (select count(1) as count from movielens_movies ) as movie_ratio
from (
  select prerule as movieid
  from pal_movielens_apriori_result
  where prerule not like '%&%'
  group by prerule
);

 * hana://ML_USER:Welcome18@hxehost:39015


movie_count movie_ratio
0          200     2.19178

Only 660 of the initial users will receive the requested 5 recommendations. However, only 200 distinct movies (2.2%) in total can be proposed overall and 41 movies (0.45%) in the top 5 lists.

## **Content-based filtering - Validate the result**
Let's verify how many movies will actually get recommendations using the following SQL:

In [15]:
%%sql
select reco_count, count(1) as movie_count
from (
  select movieid, max(rank) as reco_count
  from pal_movielens_apriori_result_contentbased
  group by movieid
) group by reco_count order by 1 desc;

 * hana://ML_USER:Welcome18@hxehost:39015


reco_count  movie_count
0           5          166
1           4            5
2           3            7
3           2            8
4           1           14

Let's verify how many distinct similar movies will actually get recommended to a movie (part of the top 5 scores) using the following SQL:

In [16]:
%%sql
select
    count(1) as movie_count
  , count(1) *100 / (select count(1) as count from movielens_movies ) as movie_ratio
from (
  select movieid
  from pal_movielens_apriori_result_contentbased
  group by movieid
);

 * hana://ML_USER:Welcome18@hxehost:39015


movie_count movie_ratio
0          200     2.19178

Let's verify how many rating does the movies with no recommendation have using the following SQL:

In [17]:
%%sql
select rating_count, count(1) as movie_count
from (
  select ratings.movieid, count(1) as rating_count
  from movielens_ratings ratings
  left outer join (
    select movieid
    from (
      select prerule as movieid
      from pal_movielens_apriori_result
      where prerule not like '%&%'
      group by prerule
    )
  ) t1 on (ratings.movieid = t1.movieid)
  where t1.movieid is null
  group by ratings.movieid
) group by rating_count;

 * hana://ML_USER:Welcome18@hxehost:39015


rating_count  movie_count
0              1         3063
1              2         1202
2              3          755
3              4          550
4              5          397
5              6          305
6              7          206
7              8          187
8              9          156
9             10          162
10            11          133
11            12          101
12            13          112
13            14           98
14            15           73
15            16           68
16            17           70
17            18           60
18            19           65
19            20           56
20            21           62
21            22           65
22            23           54
23            24           37
24            25           34
25            26           45
26            27           34
27            28           33
28            29           35
29            30           27
30            31           32
31            32           24
32            33           23
33            34           20
34            35           26
35            36           17
36            37           25
37            38           20
38            39           18
39            40           20
40            41            8
41            42           22
42            43           17
43            44           17
44            45           19
45            46           14
46            47           12
47            48           16
48            49           18
49            50            9
50            51           22
51            52           14
52            53           11
53            54            6
54            55           10
55            56            6
56            57           11
57            58            8
58            59           13
59            60            6
60            61            9
61            62           11
62            63            8
63            64            8
64            65            8
65            66            5
66            67            6
67            68            4
68            69            6
69            70            8
70            71            5
71            72            5
72            73            5
73            74            7
74            75            5
75            76            8
76            77            1
77            78            5
78            79            4
79            80            6
80            81            2
81            82            1
82            83            1
83            84            2
84            85            2
85            86            1
86            87            2
87            89            1
88            92            1

As you can see, the movies with no recommendations have up to 92 ratings, and this list include the 3063 movies with only one rating and the 1202 with only 2 ratings.